lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp

```diff
@@ -383,50 +700,123 @@ void IMConstPropPass::markBlockExecutable(Block *block) {
+  auto &destTypes = getLeafGroundTypes(regReset.getType());
+
+  // Iterate over each ground type and merge lattice values separatetly.
+  for (unsigned i = 0, e = destTypes.size(); i < e; i++) {
+    auto destType = destTypes[i].cast<FIRRTLType>();
+    auto srcValue =
+        getExtendedLatticeValue({regReset.resetValue(), i}, destType,
+                                /*allowTruncation=*/true);
+    mergeLatticeValue({regReset, i, /*isKnownRoot=*/true}, srcValue);
+  }
+}
+
+void IMConstPropPass::visitRegResetOp(RegResetOp regReset,
+                                      ValueAndLeafIndex changedValue) {
+  auto [srcRoot, srcIndex, srcSize] =
+      getRootValueWithCorrespondingChildIndexRange(regReset.resetValue());
+
+  // If source is not changed value, we don't have to process this operation.
+  if (srcRoot != changedValue.getValue())
+    return;
+
+  // If chandedValue is not included in the range of the source value, just
+  // skip.
+  if (srcIndex > changedValue.getLeafIndex() ||
+      srcIndex + srcSize <= changedValue.getLeafIndex())
+    return;
+
+  unsigned index = changedValue.getLeafIndex() - srcIndex;
+
+  auto &destTypes = getLeafGroundTypes(regReset.getType());
+  auto srcValue = getExtendedLatticeValue({regReset.resetValue(), index},
+                                          destTypes[index].cast<FIRRTLType>(),
                                           /*allowTruncation=*/true);
-  mergeLatticeValue(regReset, srcValue);
+  mergeLatticeValue({regReset, index, /*isKnownRoot=*/true}, srcValue);
 }
 
 void IMConstPropPass::markMemOp(MemOp mem) {
   for (auto result : mem.getResults())
     markOverdefined(result);
 }
 
+void IMConstPropPass::markSubindexOp(SubindexOp subindex) {
+  auto offset =
+      getChildIndexOffset(subindex.input().getType(), subindex.index());
+
+  auto rootAndIndex = translateToRootIndex({subindex.input(), offset});
+
+  // Register the current position to `valueToRootValueAndLeafIndex`.
+  valueToRootValueAndLeafIndex[subindex] = rootAndIndex;
+
+  for (unsigned i = 0; i < getNumberOfLeafs(subindex.getType()); i++)
+    rootToChildrenAccess[{rootAndIndex.getValue(),
+                          rootAndIndex.getLeafIndex() + i,
+                          /*isKnownRoot=*/true}]
+        .push_back(subindex);
+}
+
+void IMConstPropPass::markSubfieldOp(SubfieldOp subfield) {
+  auto offset =
+      getChildIndexOffset(subfield.input().getType(), subfield.fieldIndex());
+
+  auto rootAndIndex = translateToRootIndex({subfield.input(), offset});
+
+  // Register the current position to `valueToRootValueAndLeafIndex`.
+  valueToRootValueAndLeafIndex[subfield] = rootAndIndex;
+
+  for (unsigned i = 0; i < getNumberOfLeafs(subfield.getType()); i++)
+    rootToChildrenAccess[{rootAndIndex.getValue(),
+                          rootAndIndex.getLeafIndex() + i,
+                          /*isKnownRoot=*/true}]
+        .push_back(subfield);
+}
```

Comment:
These methods could be combined to save on the duplication (with an internal switch to choose between `.fieldIndex()` and `.index()`.

Reply:
Thanks, I combined them